components/script/dom/console.rs

```diff
@@ -65,13 +73,80 @@ fn console_message(global: &GlobalScope, message: DOMString, level: LogLevel) {
     })
 }
 
+//generate matrix of table elements
+#[allow(unsafe_code)]
+unsafe fn generate_table(
+    cx: &SafeJSContext,
+    tabular_data: &HandleValue,
+    _properties: &Option<Vec<DOMString>>,
+) -> Option<Vec<Vec<DOMString>>> {
+    let cx_ptr = *(cx.deref()) as *mut JSContext;
+
+    if !tabular_data.get().is_object() {
+        return None;
+    }
+
+    if is_array_like(cx_ptr, *tabular_data) {
+        let vec: Vec<DOMString> =
+            Vec::<DOMString>::from_jsval(cx_ptr, *tabular_data, StringificationBehavior::Empty)
+                .expect("Error when converting Array object to Vec<DOMString>")
+                .get_success_value()
+                .unwrap_or(&Vec::new())
+                .to_vec();
+
+        let mut res: Vec<Vec<DOMString>> = Vec::new();
+        res.push(vec!["(index)".into(), "Values".into()]);
+        for (index, word) in vec.into_iter().enumerate() {
+            res.push(vec![index.to_string().into(), word]);
+        }
+
+        return Some(res);
+    }
+
+    None
+}
+
+//transform matrix of table elements into DOMString for logging
+fn generate_table_as_string(table: &[Vec<DOMString>]) -> DOMString {
+    let mut res: DOMString = DOMString::new();
+    for row in table.iter() {
+        for col in row.iter() {
+            res.push_str(col as &str);
+            res.push_str("\t\t");
+        }
+
+        res.push_str("\n");
+    }
+
+    res
+}
+
 #[allow(non_snake_case)]
 impl Console {
     // https://developer.mozilla.org/en-US/docs/Web/API/Console/log
     pub fn Log(global: &GlobalScope, messages: Vec<DOMString>) {
         console_messages(global, &messages, LogLevel::Log)
     }
 
+    //https://console.spec.whatwg.org/#table
+    #[allow(unsafe_code)]
+    pub fn Table(
+        cx: SafeJSContext,
+        global: &GlobalScope,
+        tabular_data: HandleValue,
+        properties: Option<Vec<DOMString>>,
+    ) {
+        unsafe {
```

Comment:
This `unsafe` block is very wide. Try to move it down to exactly the problematic code and then let's see if we can eliminate it entirely.

Reply:
The problematic part is calling the generate_table() function, which is an unsafe function because it calls Vec<T>::from_jsval() and is_array_like(). I'm not very experienced with the servo documentation, so there could be a way to implement generate_table() safely which I'm not aware of.

Reply:
At the very least here, I think you should reduce the unsafe block to enclose only the unsafe operations instead of marking the entire function as unsafe.

Reply:
Okay, sure! I've reduced the unsafe block to only the call to generate_table().

Reply:
Actually, I've removed the unsafe blocked around generate_table() and just restricted the unsafe blocks to the unsafe operations inside the function, like you said.